File: tests/zig/exp.zig
const std = @import("std");
const profile = @import("profile.zig");

var result: f32 = 0;
var count: u32 = 1_000_000;

noinline fn expLoop(n: u32) f32 {
    var sum: f32 = 0;
    var i: u32 = 0;
    while (i < n) : (i += 1) {
        sum += @exp(1.0 / (1.0 + @as(f32, @floatFromInt(i))));
    }
    return sum;
}

fn run(_: void) void {
    result = expLoop(@as(*volatile u32, &count).*);
    std.mem.doNotOptimizeAway(result);
}

pub fn main(init: std.process.Init) !void {
    try profile.profile(init, "exp loop", run, {});
    // f32 accumulation: past 2^20 every term rounds to 1.0, so the sum lands within a few ulps of 1e6
    if (@abs(result - 1.0e6) >= 16.0) profile.fail("exp loop: sum is off");
}

File: tests/cpp/exp.cpp
#include "profile.h"
#include <cmath>

// every kernel takes its size through a volatile read, so the optimizer makes a real call per
// repetition instead of hoisting a pure call out of the harness loop
static volatile int count = 1000000;

PROFILE_NOINLINE float expLoop(int n) {
    float ret = 0;
    for (int i = 0; i < n; ++i) ret += expf(1.f / (1.f + i));
    return ret;
}

int main() {
    float result = 0;
    profile::run("exp loop", [&] { result = expLoop(count); profile::keep(result); });
    // f32 accumulation: past 2^20 every term rounds to 1.0, so the sum lands within a few ulps of 1e6
    if (fabsf(result - 1e6f) >= 16.f) profile::fail("exp loop: sum is off");
    return 0;
}
